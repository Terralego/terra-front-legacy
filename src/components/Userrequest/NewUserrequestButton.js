import React from 'react';
import { Link } from 'react-router-dom';
import { Button } from 'antd';

const NewUserrequestButton = ({ type = 'primary', className, style }) => (
  <Link to="/new-request">
    <Button
      icon="file-add"
      type={type}
      className={className}
      style={style}
    >
      Déclarer une activité
    </Button>
  </Link>
);

export default NewUserrequestButton;
