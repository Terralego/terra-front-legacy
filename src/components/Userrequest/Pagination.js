import React from 'react';
import { withRouter } from 'react-router-dom';
import { Pagination as AntPagination } from 'antd';
import settings from 'front-settings';

import styles from './UserrequestList.module.scss';

class Pagination extends React.Component {
  handlePaginationChange = (page, pageSize) => {
    this.props.history.push(`/manage-request/?limit=${pageSize}&page=${page}`);
  }

  render () {
    const { params, count } = this.props;

    return (
      <AntPagination
        defaultCurrent={1}
        current={params.page}
        className={styles.pagination}
        defaultPageSize={settings.PAGE_SIZE}
        pageSize={params.limit || settings.PAGE_SIZE}
        showTotal={total => `Total ${total} demandes`}
        total={count}
        onChange={this.handlePaginationChange}
      />
    );
  }
}

export default withRouter(Pagination);
