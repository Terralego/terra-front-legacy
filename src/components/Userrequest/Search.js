import React from 'react';
import { withRouter } from 'react-router-dom';
import { Input } from 'antd';
import queryString from 'query-string';

class Search extends React.Component {
  getDefaultValue = () => queryString.parse(this.props.location.search).search;

  handleSearchChange = search => {
    this.props.handleQueryUpdate({ search }, true);
  }

  render () {
    return (
      <Input.Search
        addonBefore={window.screen.width <= 400 ? '' : 'Chercher :'}
        placeholder="Numéro de déclaration, titre de l'événement&hellip;"
        onSearch={this.handleSearchChange}
        defaultValue={this.getDefaultValue()}
        enterButton
      />
    );
  }
}

export default withRouter(Search);
