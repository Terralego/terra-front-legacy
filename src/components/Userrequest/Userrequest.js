import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect, withRouter } from 'react-router-dom';
import { Spin, Row, Col, Card } from 'antd';

import { fetchUserrequest } from 'modules/userrequestList';
import { openDraft } from 'modules/userrequest';
import { getUserGroup } from 'modules/authentication';
import Summary from 'components/Summary/Summary';
import RequestStatus from 'components/RequestStatus/RequestStatus';
import Comments from 'components/Comments/Comments';
import Form from 'components/Form/Form';

const DRAFT_STATUS = 100;

class Userrequest extends React.Component {
  componentDidMount () {
    if (this.props.data && this.props.data.state === DRAFT_STATUS) {
      this.props.openDraft(this.props.data);
    }
    if (!this.props.data && !this.props.loading) {
      this.props.fetchUserrequest(this.props.match.params.id);
    }
  }

  render () {
    const { data, loading } = this.props;
    if (loading) {
      return <Spin style={{ margin: '30px auto', width: '100%' }} />;
    }
    if (data) {
      if (data.error && data.error.status === 404) {
        return <Redirect to="/manage-request" from={this.props.location.pathname} />;
      }
      // If userrequest is in draft status
      // and user group is user, redirect to editable request
      if (data.state === DRAFT_STATUS
        && this.props.userGroup === 'user') {
        return (
          <Form {...this.props} />
        );
      }
    }
    return (
      <Row gutter={24} style={{ paddingBottom: 24 }}>
        <Col span={24} lg={14}>
          {data ?
            <Summary data={data} />
          : <Spin style={{ margin: '30px auto', width: '100%' }} />}
        </Col>

        <Col span={24} lg={10}>
          {data && <RequestStatus userrequest={data} />}
          <Card title="Ã‰changes" style={{ marginTop: 24 }}>
            <Comments userrequestId={this.props.match.params.id} />
          </Card>
        </Col>
      </Row>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  // TODO: use Reselect for increase performances
  data: state.userrequestList.items[ownProps.match.params.id],
  loading: state.userrequestList.loading,
  userGroup: getUserGroup(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ fetchUserrequest, openDraft }, dispatch);

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Userrequest));
