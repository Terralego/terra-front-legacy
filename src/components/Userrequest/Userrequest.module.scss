
.userrequest {
  padding-bottom: 24px;
  &_comments {
    :global(.ant-alert) {
      margin-bottom: 24px;
    }
    :global(.ant-card) {
      margin-bottom: 24px;
    }
  }
}

@media print {
  .userrequest {
    :global(.ant-card) {
      break-inside: avoid-page;
    }
    &_summary {
      width: 100%;
    }
    &_comments {
      display: flex;
      width: 100%;
      flex-direction: column-reverse;
    }
  }
}