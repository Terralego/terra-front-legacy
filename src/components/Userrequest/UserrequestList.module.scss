.header {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  > * {
    margin-bottom: 24px;
  }
};

.actions {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  &__button {
    margin-bottom: 12px;
    margin-left: 12px;
  }
}

.pagination {
  text-align: right;
  margin-top: 12px;
  clear: both;
}

@media (max-width: 768px) {
  .actions {
    display: block;
    &__button {
      margin-bottom: 12px;
      margin-left: 0;
      margin-right: 12px;
    }
  }
}
