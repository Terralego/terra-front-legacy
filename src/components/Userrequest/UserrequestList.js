import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter, Redirect } from 'react-router-dom';
import { Table, Icon, Modal, Button, message } from 'antd';
import queryString from 'query-string';

import { getUserGroup } from 'modules/authentication';
import { submitData, saveDraft } from 'modules/userrequest';
import { requestUserrequestPage, updateState, getUserrequestsArrayFilteredByUser } from 'modules/userrequestList';
import { getPaginationParams, isCurrentPageFetching, resetPaginationCache } from 'modules/pagination';

import getColumns from 'helpers/userrequestListColumns';

import NewUserrequestButton from 'components/Userrequest/NewUserrequestButton';
import Pagination from 'components/Userrequest/Pagination';
import Search from 'components/Userrequest/Search';

import styles from './UserrequestList.module.scss';

class UserrequestList extends React.Component {
  state = {
    selectedRowKeys: [],
  };

  componentDidMount () {
    if (!this.props.loading) {
      this.props.requestUserrequestPage(this.props.location.search);
    }
  }

  componentDidUpdate (prevProps) {
    if (prevProps.location.search !== this.props.location.search && !this.props.loading) {
      this.props.requestUserrequestPage(this.props.location.search);
    }

    if (prevProps.draft.id !== this.props.draft.id) {
      this.props.resetPaginationCache('/userrequest/');
      message.destroy();
    }
  }

  onSelectChange = selectedRowKeys => {
    this.setState({ selectedRowKeys });
  }

  /**
   * getSelectedItems
   * @param  {array} properties : this.props.items
   * @return {array} properties : clone of this.props.items filter by selected items
   */
  getSelectedItems = () => (
    this.props.items
      .filter(item => this.state.selectedRowKeys.includes(item.id))
      .map(item => ({
        ...item,
        properties: {
          ...item.properties,
        },
      }))
  )

  /**
   * queryUpdate add query string parameter in url
   * duplicate function with Search component : see https://github.com/supasate/connected-react-router
   * to implement history change in actions / reducers
   * @param {object} query : couple(s) of key / value parameter(s)
   */
  handleQueryUpdate = (query, reset) => {
    const params = { ...query };
    if (reset) {
      this.props.resetPaginationCache('/userrequest/');
      params.page = 1;
    }
    return this.props.history.push(`/manage-request/?${
      queryString.stringify({
        ...queryString.parse(this.props.location.search),
        ...params,
      })
    }`);
  }

  handleCopy = () => {
    const selectedItems = this.getSelectedItems();
    Modal.confirm({
      title: selectedItems.length > 1 ? 'Êtes-vous sûr de vouloir dupliquer ces demandes ?' : 'Êtes-vous sûr de vouloir dupliquer cette demande ?',
      content: selectedItems.length > 1 ? 'Les nouvelles demandes prendront le statut "Brouillon".' : 'La nouvelle demande prendra le statut "Brouillon".',
      onOk: () => {
        message.loading('Duplication de la demande en cours...', 2.5);
        const item = selectedItems[0];
        delete item.id;
        item.properties.title += ' - copie';
        this.props.saveDraft(item);
        this.setState({
          selectedRowKeys: [],
        });
      },
    });
  }

  handleCancel = () => {
    const selectedItems = this.getSelectedItems();
    Modal.confirm({
      title: selectedItems.length > 1 ? 'Êtes-vous sûr de vouloir annuler ces demandes ?' : 'Êtes-vous sûr de vouloir annuler cette demande ?',
      content: selectedItems.length > 1 ? 'Les nouvelles demandes prendront le statut "Annuler".' : 'La nouvelle demande prendra le statut "Annuler".',
      onOk: () => {
        selectedItems
          .forEach(item => {
            this.props.updateState(item.id, -2);
          });
        this.setState({
          selectedRowKeys: [],
        });
      },
    });
  }

  handleClickOnRow (id) {
    this.props.history.push(`/manage-request/detail/${id}`);
  }

  render () {
    const { selectedRowKeys } = this.state;
    const { draft, location, userGroup, columns, pagination } = this.props;

    // If a draft newly created, redirect on its
    if (draft.id) {
      return (
        <Redirect
          to={{ pathname: `/manage-request/detail/${draft.id}`,
          state: {
            from: `${location.pathname}${location.search}`,
          } }}
        />
      );
    }

    const rowSelection = {
      selectedRowKeys,
      onChange: this.onSelectChange,
    };
    const hasSelected = selectedRowKeys.length > 0;

    return (
      <div>
        <div className={styles.header}>
          <h1 className={styles.header__title}>Déclarations d&apos;activité(s)</h1>
          <NewUserrequestButton className={styles.header__button} />
        </div>
        <div className={styles.header}>
          <Search handleQueryUpdate={this.handleQueryUpdate} />
        </div>
        {(userGroup !== 'N1' && userGroup !== 'N2') && (
          <div className={styles.actions}>
            <Button
              className={styles.actions__button}
              onClick={this.handleCopy}
              disabled={!hasSelected || selectedRowKeys.length > 1}
            >
              <Icon type="copy" />
              Dupliquer la demande
            </Button>
            <Button
              className={styles.actions__button}
              type="danger"
              onClick={this.handleCancel}
              disabled={!hasSelected}
            >
              <Icon type="minus-circle-o" className={styles.actions__iconCancel} />
              Annuler {selectedRowKeys.length} {selectedRowKeys.length > 1 ? 'demandes sélectionnées' : 'demande sélectionnée'}
            </Button>
          </div>
        )}
        <Table
          rowKey="id"
          columns={columns}
          dataSource={this.props.items}
          rowSelection={(userGroup !== 'N1' && userGroup !== 'N2') ? rowSelection : null}
          loading={this.props.loading}
          onRow={record => (
            {
              onClick: () => {
                this.handleClickOnRow(record.id);
              },
            }
          )}
          pagination={false}
        />
        <Pagination {...pagination} handleQueryUpdate={this.handleQueryUpdate} />
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  draft: state.userrequest,
  items: getUserrequestsArrayFilteredByUser(state, ownProps.location.search),
  loading: isCurrentPageFetching(state.pagination.userrequestList, ownProps.location.search),
  userGroup: getUserGroup(state),
  columns: getColumns(getUserGroup(state)),
  pagination: getPaginationParams(state.pagination.userrequestList, ownProps.location.search),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({
    requestUserrequestPage,
    resetPaginationCache,
    submitData,
    saveDraft,
    updateState,
  }, dispatch);

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(UserrequestList));
