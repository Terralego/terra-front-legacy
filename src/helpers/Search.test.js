import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Search } from 'components/Userrequest/Search';

Enzyme.configure({ adapter: new Adapter() });

function setup () {
  const props = {
    handleQueryUpdate: jest.fn(),
    location: { search: 'Hadooooken' },
  };

  const enzymeWrapper = shallow(<Search {...props} />);

  return {
    props,
    enzymeWrapper,
  };
}

describe('components', () => {
  describe('Search', () => {
    it('should render', () => {
      const tree = renderer
        .create(<Search handleQueryUpdate={jest.fn()} location={{ search: '' }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should have default values', () => {
      const { enzymeWrapper, props } = setup();
      const searchProps = enzymeWrapper.find('Search').props();
      const {
        addonBefore,
        placeholder,
        enterButton,
        onSearch,
      } = searchProps;

      expect(addonBefore).toEqual('');
      expect(placeholder).toEqual("Numéro de déclaration, titre de l'événement…");
      expect(enterButton).toBe(true);
      onSearch('Hadoken');
      expect(props.handleQueryUpdate.mock.calls).toEqual([[{ search: 'Hadoken' }, true]]);
    });
  });
});
