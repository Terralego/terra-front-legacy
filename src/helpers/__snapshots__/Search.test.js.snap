// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Search should render 1`] = `
<span
  className="ant-input-search ant-input-search-enter-button ant-input-affix-wrapper"
  style={undefined}
>
  <input
    className="ant-input"
    defaultValue={undefined}
    disabled={false}
    onKeyDown={[Function]}
    placeholder="Numéro de déclaration, titre de l'événement…"
    size={undefined}
    style={null}
    type="text"
  />
  <span
    className="ant-input-suffix"
  >
    <button
      className="ant-btn ant-input-search-button ant-btn-primary"
      disabled={undefined}
      onClick={[Function]}
      type="button"
    >
      <i
        className="anticon anticon-search"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="search"
          fill="currentColor"
          height="1em"
          onClick={undefined}
          style={undefined}
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M608 112c-167.9 0-304 136.1-304 304 0 70.3 23.9 135 63.9 186.5L114.3 856.1c-3.1 3.1-3.1 8.2 0 11.3l42.3 42.3c3.1 3.1 8.2 3.1 11.3 0l253.6-253.6C473 696.1 537.7 720 608 720c167.9 0 304-136.1 304-304S775.9 112 608 112zm161.2 465.2C726.2 620.3 668.9 644 608 644s-118.2-23.7-161.2-66.8C403.7 534.2 380 476.9 380 416s23.7-118.2 66.8-161.2c43-43.1 100.3-66.8 161.2-66.8s118.2 23.7 161.2 66.8c43.1 43 66.8 100.3 66.8 161.2s-23.7 118.2-66.8 161.2z"
          />
        </svg>
      </i>
    </button>
  </span>
</span>
`;
